src/uct/ib/dc/dc_mlx5_ep.h

```diff
@@ -298,13 +300,69 @@ uct_dc_mlx5_ep_from_dci(uct_dc_mlx5_iface_t *iface, uint8_t dci_index)
     /* Can be used with dcs* policies only, with rand policy every dci may
      * be used by many eps */
     ucs_assert(!uct_dc_mlx5_iface_is_dci_shared(iface));
-    return iface->tx.dcis[dci_index].ep;
+    return ucs_array_elem(&iface->tx.dcis, dci_index).ep;
 }
 
 void uct_dc_mlx5_ep_handle_failure(uct_dc_mlx5_ep_t *ep,
                                    struct mlx5_cqe64 *cqe,
                                    ucs_status_t status);
 
+static UCS_F_ALWAYS_INLINE void
+uct_dc_mlx5_init_dci_config_key(uct_dc_mlx5_dci_config_t *dci_config,
+                                uint8_t path_index,
+                                uint8_t is_keepalive,
+                                uint8_t max_rd_atomic)
+{
+    dci_config->key.path_index = path_index;
+    dci_config->key.flags      = 0;
+
+    if (is_keepalive) {
+        dci_config->key.flags |= UCT_DC_MLX5_DCI_CONFIG_KEEPALIVE;
+    }
+
+    if (max_rd_atomic == 64) {
+        dci_config->key.flags |= UCT_DC_MLX5_DCI_CONFIG_MAX_RD_ATOMIC_IS_64;
+    }
+
+    memset(dci_config->key.padding, 0, sizeof(dci_config->key.padding));
+}
+
+ucs_status_t static UCS_F_ALWAYS_INLINE
+uct_dc_mlx5_dci_pool_add_dci(uct_dc_mlx5_iface_t *iface, uint8_t pool_index,
+                             uint8_t dci_index)
+{
+    uct_dc_mlx5_dci_pool_t *pool = &iface->tx.dci_pool[pool_index];
+    uct_dc_mlx5_dci_config_t pool_config = {
+        .u64 = pool->config_key
+    };
+    ucs_status_t status;
+    uint8_t* stack_ptr;
+
+    ucs_assertv(ucs_array_length(&pool->stack) <= iface->tx.ndci,
+                "stack length exceeded ndci");
+
+    if ((ucs_array_capacity(&pool->stack) == ucs_array_length(&pool->stack)) ||
```

Comment:
this looks weird

Reply:
This means all dcis qps are already allocated, no?